### DISCUSSION TITLE
Listing web-mqtt connections fails with `:not_found`

### DISCUSSION THREAD

**Discussion by johanrhodin** [rabbitmq/rabbitmq-server#9302] [Other] [answered]
Listing Web-MQTT connections in the HTTP API work, but when trying to list them with `rabbitmqctl list_mqtt_connections` they show up as `:not_found`:
```
rabbitmqctl list_mqtt_connections
Listing MQTT connections ...
:not_found
```

Compare to HTTP API:
```
  {
    "auth_mechanism": "none",
    "channel_max": 0,
    "client_properties": {
      "client_id": "mqttjs_ada70c"
    },
    "connected_at": 1693942464196,
    "frame_max": 0,
    "garbage_collection": {
      "fullsweep_after": 65535,
      "max_heap_size": 0,
      "min_bin_vheap_size": 46422,
      "min_heap_size": 233,
      "minor_gcs": 3
    },
    "host": "127.0.0.1",
    "name": "127.0.0.1:48384 -> 127.0.0.1:15675",
    "node": "rabbit@test-tall-gold-lark-01",
    "peer_cert_issuer": null,
    "peer_cert_subject": null,
    "peer_cert_validity": null,
    "peer_host": "127.0.0.1",
    "peer_port": 48384,
    "port": 15675,
    "protocol": "Web MQTT 3-1-1",
    "recv_cnt": 12,
    "recv_oct": 787,
    "recv_oct_details": {
      "rate": 0
    },
    "reductions": 86428,
    "reductions_details": {
      "rate": 129.6
    },
    "send_cnt": 12,
    "send_oct": 356,
    "send_oct_details": {
      "rate": 0
    },
    "send_pend": 0,
    "ssl": false,
    "ssl_cipher": null,
    "ssl_hash": null,
    "ssl_key_exchange": null,
    "ssl_protocol": null,
    "state": "running",
    "timeout": 30,
    "user": "rwgvqgbl",
    "user_who_performed_action": "rwgvqgbl",
    "vhost": "rwgvqgbl"
  }
```

This is RabbitMQ `3.12.4` with Erlang `25.3`, but this issue is present in older (pre 3.12) versions too.

**Comment by gomoripeti** [comment, marked as answer]
the issue is that the `list_mqtt_connections` command
- calls for all `local_connection_pids()` `rabbit_mqtt_reader:info/2` https://github.com/rabbitmq/rabbitmq-server/blob/main/deps/rabbitmq_mqtt/src/rabbit_mqtt.erl#L72
- which does a `gen_server:call`. https://github.com/rabbitmq/rabbitmq-server/blob/main/deps/rabbitmq_mqtt/src/rabbit_mqtt_reader.erl#L63
- However the web-mqtt connection processes are not mqtt_readers but cowboy_websocket loops. 
- cowboy websocket loop assumes all gen_server calls are supervisor calls https://github.com/ninenines/cowboy/blob/master/src/cowboy_websocket.erl#L427
- and returns `{error, not_found}` for any other call (like the info call) https://github.com/ninenines/cowboy/blob/master/src/cowboy_children.erl#L174-L178

I'm not sure how to fully solve this, `rabbit_web_mqtt_handler` could expose an `info/2` function which would use message passing other than a `$gen_call` which would reach the [websocket_info](https://github.com/rabbitmq/rabbitmq-server/blob/main/deps/rabbitmq_web_mqtt/src/rabbit_web_mqtt_handler.erl#L146) callback. But I don't want the rabbitmq_mqtt plugin to depend on the rabbitmq_web_mqtt plugin by calling one of its functions. Also I don't know what would be the ideal way to differentiate between mqtt and web-mqtt connection processes (apart from the later returning not_found for `rabbit_mqtt_reader:info/2`)

(The management API uses completely different path, it does not query the connection process directly but reads info stored in [mgmt db](https://github.com/rabbitmq/rabbitmq-server/blob/main/deps/rabbitmq_management/src/rabbit_mgmt_wm_connection.erl#L85)/core metrics table which is populated by the connection processes by [emitting stats](https://github.com/rabbitmq/rabbitmq-server/blob/main/deps/rabbitmq_web_mqtt/src/rabbit_web_mqtt_handler.erl#L244)).

**Comment by gomoripeti** [comment]
Opened a draft PR with a proposal that has the problem of mqtt plugin calling web_mqtt plugin https://github.com/rabbitmq/rabbitmq-server/pull/9697
